### PR TITLE
Move the testing MockProvider into a dedicated shared package

### PR DESCRIPTION
**Pull request by liamcervante**
**This is an entirely internal change that should not affect any external users.**

I've followed up on a FIXME that I saw in the comments, and realised it was something that's also been annoying me for a while. This PR moves the `MockProvider` that was in the `terraform` package into the a new `providers/testing` package, given it is now used from many different parts of Terrform Core including the `terraform` package, but also the `command` package, and now the Stacks implementation.

Note, there is already a mock provider within the `providers` package, but this is used by the testing framework and is actually a feature that is available to external users. So I put this internal testing mock provider into the `testing` subpackage, to make it clear that one is an actual published Terraform feature and one is designed to make testing easier.

### CODE REVIEW

File: internal/terraform/resource_provider_mock_test.go [hunk 1]
```diff
@@ -80,10 +82,19 @@ func simpleMockProvider() *MockProvider {
 	}
 }
 
-// ProviderSchema is a helper to convert from the internal GetProviderSchemaResponse to
-// a ProviderSchema.
-func (p *MockProvider) ProviderSchema() *ProviderSchema {
-	resp := p.getProviderSchema()
+// ProviderSchema is a helper to convert from the internal
```
Comment (alisdair) [inline internal/terraform/resource_provider_mock_test.go:85]: This comment doesn't match the function name below.

File: internal/terraform/resource_provider_mock_test.go [hunk 2]
```diff
@@ -80,10 +82,19 @@ func simpleMockProvider() *MockProvider {
...
+// GetProviderSchemaResponse to a ProviderSchema.
+func GetProviderSchema(p *testing_provider.MockProvider) *ProviderSchema {
```
Comment (alisdair) [inline internal/terraform/resource_provider_mock_test.go:87]: Can we make this an unexported function? And while we're doing this refactor, maybe we can also make `ProviderSchema` unexported?

**Review by alisdair — APPROVED** [review]

File: internal/terraform/resource_provider_mock_test.go (hunk 2 above)
Comment (liamcervante) [reply internal/terraform/resource_provider_mock_test.go]: Done!

File: internal/terraform/resource_provider_mock_test.go (hunk 1 above)
Comment (liamcervante) [reply internal/terraform/resource_provider_mock_test.go]: Done!

**Pull request merged**